logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/backup/CreateBackupUseCase.kt

```diff
@@ -0,0 +1,131 @@
+}
+
+internal class CreateBackupUseCaseImpl(
+    private val userId: UserId,
+    private val getCurrentClientId: ObserveCurrentClientIdUseCase,
+    private val kaliumFileSystem: KaliumFileSystem,
+    private val dispatchers: KaliumDispatcher = KaliumDispatcherImpl,
+) : CreateBackupUseCase {
+
+    val idMapper = MapperProvider.idMapper()
+
+    override suspend operator fun invoke(password: String): CreateBackupResult = withContext(dispatchers.io) {
+        createBackupFile(userId).fold(
+            { error -> CreateBackupResult.Failure(error) },
+            { backupFilePath ->
+                val isBackupEncrypted = password.isNotEmpty()
+                var encryptedDataSize = 0L
+                var finalBackupFilePath = backupFilePath
+
+                if (isBackupEncrypted) {
+                    val encryptedBackupFilePath = kaliumFileSystem.tempFilePath(BACKUP_ENCRYPTED_FILE_NAME)
+                    encryptedDataSize = encryptBackup(
+                        kaliumFileSystem.source(backupFilePath),
+                        kaliumFileSystem.sink(encryptedBackupFilePath),
+                        BackupCoder.Passphrase(password)
+                    )
+                    finalBackupFilePath = "$encryptedBackupFilePath.zip".toPath()
+
+                    createCompressedFile(
+                        listOf(kaliumFileSystem.source(encryptedBackupFilePath) to encryptedBackupFilePath.name),
+                        kaliumFileSystem.sink(finalBackupFilePath)
+                    ).onFailure {
+                        return@withContext CreateBackupResult.Failure(
+                            StorageFailure.Generic(RuntimeException("Failed to compress encrypted backup file"))
+                        )
+                    }
+
+                    // Delete the temporary files
+                    kaliumFileSystem.delete(backupFilePath)
+                    kaliumFileSystem.delete(encryptedBackupFilePath)
+                    kaliumFileSystem.delete(kaliumFileSystem.tempFilePath(BACKUP_METADATA_FILE_NAME))
+
+                    if (encryptedDataSize > 0) {
+                        CreateBackupResult.Success(finalBackupFilePath, encryptedDataSize, finalBackupFilePath.name)
+                    } else {
+                        CreateBackupResult.Failure(StorageFailure.Generic(RuntimeException("Failed to encrypt backup file")))
+                    }
+                } else CreateBackupResult.Success(finalBackupFilePath, encryptedDataSize, finalBackupFilePath.name)
+            })
+    }
+
+    private suspend fun encryptBackup(backupFileSource: Source, encryptedBackupSink: Sink, passphrase: BackupCoder.Passphrase) =
+        ChaCha20Utils().encryptBackupFile(backupFileSource, encryptedBackupSink, idMapper.toCryptoModel(userId), passphrase)
+
+    private suspend fun createMetadataFile(userId: UserId): Path {
+        val clientId = getCurrentClientId.invoke().first()
```

Comment:
I think we are not really using invoke() anywhere getCurrentClientId() would be better here